internal/storage/memory.go

```diff
@@ -361,3 +387,230 @@ func inMemoryCRDB() (testserver.TestServer, error) {
+func WithHTTPClient(client *http.Client) func(svc *memoryUserInfoService) {
+	return func(svc *memoryUserInfoService) {
+		svc.httpClient = client
+	}
+}
+
+func (s *memoryUserInfoService) createTables() error {
+	_, err := s.db.Exec(`
+        CREATE TABLE IF NOT EXISTS user_info (
+            id    UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+            name  STRING NOT NULL,
+            email STRING NOT NULL,
+            sub   STRING NOT NULL,
+            iss_id   UUID NOT NULL REFERENCES issuers(id),
+            UNIQUE (iss_id, sub)
+        )`)
+
+	return err
+}
+
+// LookupUserInfoByClaims fetches UserInfo from the store.
+// This does not make an HTTP call with the subject token, so for this
+// data to be available, the data must have already be fetched and
+// stored.
+func (s memoryUserInfoService) LookupUserInfoByClaims(ctx context.Context, iss, sub string) (*types.UserInfo, error) {
+	selectCols := withQualifier([]string{
+		userInfoCols.Name,
+		userInfoCols.Email,
+		userInfoCols.Subject,
+	}, "ui")
+
+	selectCols = append(selectCols, "i."+issuerCols.URI)
+
+	selects := strings.Join(selectCols, ",")
+
+	stmt := fmt.Sprintf(`
+	SELECT %[1]s FROM user_info ui
+        JOIN issuers i ON ui.%[2]s = i.%[3]s
+        WHERE i.%[4]s = $1 and ui.%[5]s = $2`,
+		selects,
+		userInfoCols.IssuerID,
+		issuerCols.ID,
+		issuerCols.URI,
+		userInfoCols.Subject,
+	)
+	row := s.db.QueryRowContext(ctx, stmt, iss, sub)
+
+	var ui types.UserInfo
+
+	err := row.Scan(&ui.Name, &ui.Email, &ui.Subject, &ui.Issuer)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, types.ErrUserInfoNotFound
+	}
+
+	return &ui, err
+}
+
+func (s memoryUserInfoService) LookupUserInfoByID(ctx context.Context, id string) (*types.UserInfo, error) {
+	selectCols := withQualifier([]string{
+		userInfoCols.ID,
+		userInfoCols.Name,
+		userInfoCols.Email,
+		userInfoCols.Subject,
+	}, "ui")
+
+	selectCols = append(selectCols, "i."+issuerCols.URI)
+
+	selects := strings.Join(selectCols, ",")
+
+	stmt := fmt.Sprintf(`
+        SELECT %[1]s FROM user_info ui
+        JOIN issuers i ON ui.%[2]s = i.%[3]s
+        WHERE ui.%[4]s = $1
+        `, selects, userInfoCols.IssuerID, issuerCols.ID, userInfoCols.ID)
+
+	row := s.db.QueryRowContext(ctx, stmt, id)
+
+	var ui types.UserInfo
+
+	err := row.Scan(&ui.ID, &ui.Name, &ui.Email, &ui.Subject, &ui.Issuer)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, types.ErrUserInfoNotFound
+	}
+
+	return &ui, err
+}
+
+// StoreUserInfo is used to store user information by issuer and
+// subject pairs. UserInfo is unique to issuer/subject pairs.
+func (s memoryUserInfoService) StoreUserInfo(ctx context.Context, userInfo types.UserInfo) (*types.UserInfo, error) {
+	if len(userInfo.Issuer) == 0 {
+		return nil, fmt.Errorf("%w: issuer is empty", types.ErrInvalidUserInfo)
+	}
+
+	if len(userInfo.Subject) == 0 {
+		return nil, fmt.Errorf("%w: subject is empty", types.ErrInvalidUserInfo)
+	}
+
+	row := s.db.QueryRowContext(ctx, `
```

Comment:
I missed this earlier, but queries that alter DB state should really leverage transactions, which also implies userinfo and the Fosite server should use some transaction middleware.

We can probably just take the [existing middleware][storage-middleware] and move it somewhere else since things that aren't the API need it now.

[storage-middleware]: https://github.com/infratographer/identity-manager-sts/blob/d77fa3b76ee80422cd4666a1d641c89ea0188d94/internal/api/httpsrv/handler.go#L70